pkg/error_code/error_code.go

```diff
@@ -31,42 +36,117 @@
 package errcode
 
 import (
+	"fmt"
 	"net/http"
+	"strings"
 )
 
-// RegisteredCode helps document that we are using a registered error code that must never change
-type RegisteredCode string
-
-const (
-	// InternalCode means the operation placed the system is in an inconsistent or unrecoverable state.
-	// Essentially a handled panic.
-	// This is exactly the same as a HTTP 500, so it is not necessary to send this code over HTTP.
-	InternalCode RegisteredCode = "internal"
-	// InvalidInputCode is a validation failure of an input.
-	// The response will indicate the exact input issue.
-	InvalidInputCode RegisteredCode = "input"
-	// NotFoundCode indicates a resource is missing.
-	// The response can indicate what resource is missing.
-	// However, when a single resource is requested, that may be implicit.
-	NotFoundCode RegisteredCode = "missing"
+// CodeStr is a representation of the type of a particular error.
+// The underlying type is string rather than int.
+// This enhances both extensibility (avoids merge conflicts) and user-friendliness.
+// A CodeStr can have dot separators indicating a hierarchy.
+type CodeStr string
+
+func (str CodeStr) String() string { return string(str) }
+
+// A Code has a CodeStr representation.
+// It is attached to a Parent to find metadata from it.
+// The Meta field is provided for extensibility: e.g. attaching HTTP codes.
+type Code struct {
+	// CodeStr contains the full dot-separted path.
+	// This allows for equality comparison based on CodeStr.
+	CodeStr CodeStr
```

Comment:
We can store it either way in the struct regardless of whether a full or partial path is given to `Child`.
I like the perspective of correct by construction. The verbosity is a documentation feature that I felt strongly enhance readability. One can look at where the error code is declared and see the full code string rather than re-constructing it by jumping around files.
I will change the code to support both styles.

Reply:
For storage storing just the child portion as you recommend may be better. Although not currently supported and perhaps a dubious idea, it would make it easier to define children first and then define an ancestor.